migrating lazy-loading content (#3176)
GoogleChrome/web.dev: src/site/content/en/fast/lazy-loading-images/index.md

---
layout: post
title: Lazy-loading images
authors:
  - jeremywagner
  - rachelandrew
date: 2019-08-16
updated: 2020-06-09
description: |
  This post explains lazy-loading and the options available to you when lazy-loading images.
tags:
  - performance
---

Images can appear on a webpage due to being inline in the HTML as `<img>` elements
or as CSS background images. In this guide you will find out how to lazy-load both types of image.

## Inline images {: #images-inline }

The most common lazy-loading candidates are images as used in `<img>` elements.
With inline images we have three options for lazy-loading,

> which may be used in combination for the best compatibility across browsers.
> 
> - [Using native browser lazy-loading](#images-inline-native).
> - [Using Intersection Observer](#images-inline-intersection-observer).
> - [Using scroll and resize event handlers](#images-inline-event-handlers).

Reviewer: ```suggestion
which may be used in combination for the best compatibility across browsers:

- [Using native browser lazy-loading](#images-inline-native)
- [Using Intersection Observer](#images-inline-intersection-observer)
- [Using scroll and resize event handlers](#images-inline-event-handlers)
```

### Using native browser lazy-loading {: #images-inline-native }

Chrome and Firefox both support native browser lazy-loading with the `loading` attribute.
This attribute can be added to `<img>` elements, and also to `<iframe>` elements.
A value of `lazy` tells the browser to load the image immediately if it is in the viewport,
and to fetch other images when the user scrolls near them.

{% Aside %}

>   Please note `<iframe loading=lazy>` is currently non-standard.

Reviewer: ```suggestion
  Please note `<iframe loading="lazy">` is currently non-standard.
```

  While implemented in Chromium, it does not yet have a specification and is subject to future change when this does happen.

>   We suggest not to lazy-load iframes using the loading attribute until it becomes part of the specification.

Reviewer: ```suggestion
  We suggest not to lazy-load iframes using the `loading` attribute until it becomes part of the specification.
```

{% endAside %}

See the `loading` field of MDN's

> [Browser compatibility](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#Browser_compatibility)

Reviewer: ```suggestion
[browser compatibility](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#Browser_compatibility)
```

table for details of browser support for native lazy-loading.
If the browser does not support native lazy-loading then the attribute will be ignored
and images will load immediately, as normal.

For most websites, adding this attribute to inline images will be a performance boost
and save users loading images that they may not ever scroll to.
If you have large numbers of images and want to be sure that users of browsers without support for native lazy-loading benefit
you will need to combine this with one of the methods explained next.

> 

Reviewer: ```suggestion

To learn more, check out [Native lazy-loading for the web](/native-lazy-loading/).
```

      entries.forEach(function(entry) {
        if (entry.isIntersecting) {
          let lazyImage = entry.target;
          lazyImage.src = lazyImage.dataset.src;
          lazyImage.srcset = lazyImage.dataset.srcset;
          lazyImage.classList.remove("lazy");
          lazyImageObserver.unobserve(lazyImage);
        }
      });
    });

    lazyImages.forEach(function(lazyImage) {
      lazyImageObserver.observe(lazyImage);
    });
  } else {
    // Possibly fall back to event handlers here
  }
});
```

On the document's `DOMContentLoaded` event, this script queries the DOM for all
`<img>` elements with a class of `lazy`. If Intersection Observer is available,
create a new observer that runs a callback when `img.lazy` elements enter the
viewport.

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/lazy-intersection-observer?path=index.html&previewSize=0"
    title="lazy-intersection-observer on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

Intersection Observer is available in all modern browsers.

> Therefore using it as a polyfill for `loading=lazy` will ensure that lazy-loading is available for most visitors.

Reviewer: ```suggestion
Therefore using it as a polyfill for `loading="lazy"` will ensure that lazy-loading is available for most visitors.
```

is used to throttle function calls. As images are lazy-loaded, they're removed
from the elements array. When the elements array reaches a `length` of `0`, the
scroll event handler code is removed.

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/lazy-loading-fallback?path=lazy.js&previewSize=0"
    title="lazy-loading-fallback on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

While this code works in pretty much any browser, it has potential performance
issues in that repetitive `setTimeout` calls can be wasteful, even if the code
within them is throttled. In this example, a check is being run every 200
milliseconds on document scroll or window resize regardless of whether there's
an image in the viewport or not. Plus, the tedious work of tracking how many
elements are left to lazy-load and unbinding the scroll event handler are left
to the developer. You can find out more about this technique in
 [this CSS Tricks article](https://css-tricks.com/the-complete-guide-to-lazy-loading-images/#method-1-trigger-the-image-load-using-javascript-events).

> Simply put: Use native lazy-loading with a fallback intersection observer wherever possible, and only use event

Reviewer: ```suggestion
Simply put: Use native lazy-loading with a fallback Intersection Observer implementation wherever possible, and only use event
```

handlers if the widest possible compatibility is a critical application
requirement.

## Images in CSS {: #images-css }

While `<img>` tags are the most common way of using images on web pages, images
can also be invoked via the CSS
[`background-image`](https://developer.mozilla.org/en-US/docs/Web/CSS/background-image)
property (and other properties). Native browser lazy-loading does not apply to CSS background images,

> so we need to consider other methods if we have background images to lazy-load.

Reviewer: ```suggestion
so you need to consider other methods if you have background images to lazy-load.
```

`div.lazy-background` element at that time, which loads the image:

```javascript
document.addEventListener("DOMContentLoaded", function() {
  var lazyBackgrounds = [].slice.call(document.querySelectorAll(".lazy-background"));

  if ("IntersectionObserver" in window) {
    let lazyBackgroundObserver = new IntersectionObserver(function(entries, observer) {
      entries.forEach(function(entry) {
        if (entry.isIntersecting) {
          entry.target.classList.add("visible");
          lazyBackgroundObserver.unobserve(entry.target);
        }
      });
    });

    lazyBackgrounds.forEach(function(lazyBackground) {
      lazyBackgroundObserver.observe(lazyBackground);
    });
  }
});
```
<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/lazy-background?path=index.html&previewSize=0"
    title="lazy-background on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

As indicated earlier, if you need Internet Explorer support for lazy-loading of background images,
you will need to polyfill the Intersection Observer code, due to lack of support in that browser.

> ## lazy-loading libraries {: #libraries }

Reviewer: ```suggestion
## Lazy-loading libraries {: #libraries }
```

The following libraries can be used to lazy-load images.

- [lazysizes](https://github.com/aFarkas/lazysizes) is a full-featured lazy
loading library that lazy-loads images and iframes. The pattern it uses is quite
similar to the code examples shown here in that it automatically binds to a
`lazyload` class on `<img>` elements, and requires you to specify image URLs in
`data-src` and/or `data-srcset` attributes, the contents of which are swapped
into `src` and/or `srcset` attributes, respectively. It uses Intersection
Observer (which you can polyfill), and can be extended with [a number of
plugins](https://github.com/aFarkas/lazysizes#available-plugins-in-this-repo) to
do things like lazy-load video. [Find out more about using lazysizes](/use-lazysizes-to-lazyload-images/).
- [lozad.js](https://github.com/ApoorvSaxena/lozad.js) is a super lightweight
option that uses Intersection Observer only. As such, it's highly performant,
but will need to be polyfilled before you can use it on older browsers.

> - [yall.js](https://github.com/malchata/yall.js) is a library I wrote that uses

Reviewer: ```suggestion
- [yall.js](https://github.com/malchata/yall.js) is a library that uses
```

Intersection Observer and falls back to event handlers. It's compatible with IE11
and major browsers.

> - If you're seeking a React-specific lazy-loading library, you might consider

Reviewer: ```suggestion
- If you need a React-specific lazy-loading library, consider
```

---
layout: post
title: Lazy-loading images
authors:
  - jeremywagner
  - rachelandrew
date: 2019-08-16
updated: 2020-06-09
description: |
  This post explains lazy-loading and the options available to you when lazy-loading images.

> tags:
>   - performance
> ---

Reviewer: ```suggestion
tags:
  - performance
  - images
---
```

description: |
  This post explains lazy-loading and the options available to you when lazy-loading images.
tags:
  - performance
  - images
---

Images can appear on a webpage due to being inline in the HTML as `<img>` elements
or as CSS background images. In this post you will find out how to lazy-load both types of image.

## Inline images {: #images-inline }

The most common lazy-loading candidates are images as used in `<img>` elements.
With inline images we have three options for lazy-loading,
which may be used in combination for the best compatibility across browsers:

- [Using native browser lazy-loading](#images-inline-native)
- [Using Intersection Observer](#images-inline-intersection-observer)
- [Using scroll and resize event handlers](#images-inline-event-handlers)

### Using native browser lazy-loading {: #images-inline-native }

Chrome and Firefox both support native browser lazy-loading with the `loading` attribute.
This attribute can be added to `<img>` elements, and also to `<iframe>` elements.
A value of `lazy` tells the browser to load the image immediately if it is in the viewport,
and to fetch other images when the user scrolls near them.

{% Aside %}

>   Please note `<iframe loading="lazy">` is currently non-standard.

Reviewer 2: ```suggestion
  Note `<iframe loading="lazy">` is currently non-standard.
```

any of `img.lazy` elements are in the viewport. A `setTimeout` call is used to
delay processing, and an `active` variable contains the processing state which
is used to throttle function calls. As images are lazy-loaded, they're removed
from the elements array. When the elements array reaches a `length` of `0`, the
scroll event handler code is removed.

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/lazy-loading-fallback?path=lazy.js&previewSize=0"
    title="lazy-loading-fallback on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

While this code works in pretty much any browser, it has potential performance
issues in that repetitive `setTimeout` calls can be wasteful, even if the code
within them is throttled. In this example, a check is being run every 200
milliseconds on document scroll or window resize regardless of whether there's
an image in the viewport or not. Plus, the tedious work of tracking how many
elements are left to lazy-load and unbinding the scroll event handler are left
to the developer. You can find out more about this technique in

>  [this CSS Tricks article](https://css-tricks.com/the-complete-guide-to-lazy-loading-images/#method-1-trigger-the-image-load-using-javascript-events).

Reviewer 2: ```suggestion
 [The Complete Guide to Lazy Loading Images](https://css-tricks.com/the-complete-guide-to-lazy-loading-images/#method-1-trigger-the-image-load-using-javascript-events).
```

---
layout: post

> title: lazy-loading images

Reviewer: Capitalize the title and double-check other occurrences of "lazy-loading", I've noticed a few more "find and replace" misses.

Author: I found one which I think was missed due to a hard return.

authors:
  - jeremywagner
  - rachelandrew
date: 2019-08-16
updated: 2020-06-09
description: |

>   This post explains lazy-loading and the options available to you when lazy-loading images and video.

Reviewer: ```suggestion
  This post explains lazy-loading and the options available to you when lazy-loading images.
```

tags:
  - performance
---

Images can appear on a webpage due to being inline in the HTML as `<img>` elements
or as CSS background images. In this guide you will find out how to lazy-load both types of image.

> ### Inline images {: #images-inline }

Reviewer: Subheadings in this post are off by one, h3 should be h2 and so on.

table for details of browser support for native lazy-loading.
If the browser does not support native lazy-loading then the attribute will be ignored
and images will load immediately, as normal.

For most websites, adding this attribute to inline images will be a performance boost
and save users loading images that they may not ever scroll to.
If you have large numbers of images and want to be sure that users of browsers without support for native lazy-loading benefit
you will need to combine this with one of the methods explained next.

#### Using intersection observer {: #images-inline-intersection-observer }

To polyfill lazy-loading of `<img>` elements, we use JavaScript to check if they're in the
viewport. If they are, their `src` (and sometimes `srcset`) attributes are
populated with URLs to the desired image content.

If you've written lazy-loading code before, you may have accomplished your task
by using event handlers such as `scroll` or `resize`. While this approach is the
most compatible across browsers, modern browsers offer a more performant and
efficient way to do the work of checking element visibility via [the

> intersection observer API](/web/updates/2016/04/intersectionobserver).

Reviewer: Capitalize "Intersection Observer" when referring to the API and "(intersection) observer" when referring to an instance.

Author: gotcha - still checking through these

> intersection observer API](/web/updates/2016/04/intersectionobserver).

Reviewer: ```suggestion
intersection observer API](https://developers.google.com/web/updates/2016/04/intersectionobserver).
```

Reviewer: Double-check that all the links are working, I've also noticed a link that's broken because it used to be an anchor link on Web Fundamentals version.

images using this markup pattern:

```javascript
document.addEventListener("DOMContentLoaded", function() {
  var lazyImages = [].slice.call(document.querySelectorAll("img.lazy"));

  if ("IntersectionObserver" in window) {
    let lazyImageObserver = new IntersectionObserver(function(entries, observer) {
      entries.forEach(function(entry) {
        if (entry.isIntersecting) {
          let lazyImage = entry.target;
          lazyImage.src = lazyImage.dataset.src;
          lazyImage.srcset = lazyImage.dataset.srcset;
          lazyImage.classList.remove("lazy");
          lazyImageObserver.unobserve(lazyImage);
        }
      });
    });

    lazyImages.forEach(function(lazyImage) {
      lazyImageObserver.observe(lazyImage);
    });
  } else {
    // Possibly fall back to a more compatible method here
  }
});
```

On the document's `DOMContentLoaded` event, this script queries the DOM for all
`<img>` elements with a class of `lazy`. If intersection observer is available,
create a new observer that runs a callback when `img.lazy` elements enter the
viewport.

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe

>     src="https://glitch.com/embed/#!/embed/lazy-intersection-observer?path=index.html&previewSize=100"

Reviewer: I think there's an option (something you append to the Glitch URL) to choose whether the preview shows the code or the app. It's not a big deal, but I think in this case it's better to show the code by default because the app is blank initially. (And it seems like lazy-loading is triggered only when you scroll inside the Glitch iframe?)

Reviewer 3: I think setting `previewSize` to 0 will only show the code.
